### PR TITLE
[compile] allow tangents to be freed before the end of compiled bw

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #177837

Original intention:

Compiled backward holds tangents allocated by the end of backward execution.

This increases peak memory (active and reserved) for the cases of long graphs, in our case SimpleFSDP, 
This is more of a problem for big gradients, when loss function is not in the backward.
In this case gradients of last linear are proportianl to bs * seq_len * emb_dim.

Tangents are inputs of the graph and we need similar to other places "boxed" calling convention style for the inputs, to be able to be freeed after last use in compiled backward graph.

Introduce a boxed calling convention for autograd backward: PyNode::apply  moves grad tensors from the immutable args tuple into a mutable list so Python code can release individual tensor references mid-execution. For the torchtitan simple_fsdp pattern (compiled model + compiled loss), this is sufficient to free tangent memory during backward without resize_(0).

Testing:

```
  python -m pytest test/functorch/test_aotdispatch.py -k "test_tangent"
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @azahed98 @xmfan @aditvenk

### CODE REVIEW

File: torch/csrc/autograd/python_function.h [hunk 1]
```diff
@@ -120,6 +120,11 @@ struct THPFunction {
   // https://github.com/pytorch/pytorch/pull/98659#pullrequestreview-1376822560
   bool materialize_non_diff_grads;
 
+  // When true, PyNode::apply uses boxed calling convention: grad tensors are
+  // moved from the immutable args tuple into a mutable list (_boxed_grads) so
+  // the backward can release individual tensor references mid-execution.
+  bool boxed_call = false;
```
Comment (albanD) [inline torch/csrc/autograd/python_function.h]: I would expect to see proper testing of the behavior of this one outside of torch.compile

File: torch/csrc/autograd/python_function.cpp [hunk 2]
```diff
@@ -160,11 +160,36 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
 
   // Massage a C++ variable_list into a Python arguments tuple
   THPObjectPtr pyInputs(to_py_args(inputs, &_device_guard));
+  inputs.clear();
+
+  if (py_fn->boxed_call) {
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp]: Please ask claude via the pr-review skill from PT here to review this to fix refcount bugs.

File: torch/csrc/autograd/python_function.cpp [hunk 3]
```diff
@@ -1844,6 +1884,11 @@ static struct PyGetSetDef THPFunction_properties[] = {
      (setter)THPFunction_set_compiled_autograd_backward_state,
      nullptr,
      nullptr},
+    {"_boxed_call",
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp]: Can we have documentation about this?

Also this is not really boxed since you return the results normally, not via this same stack.

**Review by albanD — CHANGES_REQUESTED** [review]

File: torch/csrc/autograd/python_function.cpp (hunk 3 above)
Comment (IvanKobzarev) [reply torch/csrc/autograd/python_function.cpp]: Yes, claude suggests naming "boxed_grads_call" which is better than "boxed_call", but does not feel perfect too :) Will rename to boxed_grads_call for now.

File: torch/csrc/autograd/python_function.h (hunk 1 above)
Comment (IvanKobzarev) [reply torch/csrc/autograd/python_function.h]: Yes, will add tests in test_autograd.py

File: test/functorch/test_aotdispatch.py [hunk 4]
```diff
@@ -4733,6 +4734,201 @@ def f(x):
...
+        compiled_model = torch.compile(model, backend="inductor")
+
+        def loss_fn(pred, labels):
+            return torch.nn.functional.cross_entropy(pred.float(), labels)
+
+        compiled_loss = torch.compile(loss_fn, backend="inductor")
+
+        refcount_box = {}
+
+        class CaptureRefcount(torch.autograd.Function):
```
Comment (frgossen) [inline test/functorch/test_aotdispatch.py]: There are a few repeating things in the test cases, e.g. CaptureRefcount. Maybe move them out like you did for the model with `_make_model_and_input` so it's clearer how the test cases differ.

File: torch/_dynamo/compiled_autograd.py [hunk 5]
```diff
@@ -489,15 +489,15 @@ def call_aot_bwd_prologue(
             ctx_saved_tensors: Sequence[torch.Tensor],
             ctx_symints: Sequence[IntLikeType],
             ctx_opaque_objs: Sequence[Any],
-            *flat_args: Sequence[Any],
```
Comment (frgossen) [inline torch/_dynamo/compiled_autograd.py:494]: nit: Keep the flat_args type consistent with `_backward_prologue_functional` flat_args: Sequence[Any]

File: test/functorch/test_aotdispatch.py [hunk 6]
```diff
@@ -4733,6 +4734,201 @@ def f(x):
...
+            @staticmethod
+            def forward(ctx, pred):
+                return pred.clone()
+
+            @staticmethod
+            def backward(ctx, grad):
+                refcount_box["at_boundary"] = sys.getrefcount(grad)
```
Comment (frgossen) [inline test/functorch/test_aotdispatch.py]: Why use a dict, not a nonlocal? Does that cause graph breaks?

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 7]
```diff
@@ -2775,9 +2782,16 @@ def forward(ctx: Any, *deduped_flat_tensor_args: Any) -> Any:
 
             @staticmethod
             def backward(ctx: Any, *flat_args: Any) -> tuple[Any, ...]:
-                # Combine tensors from both sources:
-                # 1. ctx.saved_tensors - tensors that went through save_for_backward (with VC check)
-                # 2. ctx._tensors_no_vc_check - tensors stashed directly on ctx (no VC check)
```
Comment (frgossen) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: Is this comment no longer relevant? Why not move it down?

**Review by frgossen — COMMENTED** [review]
LGTM but you will want a second opinion. Left some comments.

File: test/functorch/test_aotdispatch.py (hunk 6 above)
Comment (IvanKobzarev) [reply test/functorch/test_aotdispatch.py]: No preference, just used dict for this usually :)

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 7 above)
Comment (IvanKobzarev) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: Relevant, will return it back :)

File: torch/csrc/autograd/python_function.h [hunk 8]
```diff
@@ -120,6 +120,11 @@ struct THPFunction {
   // https://github.com/pytorch/pytorch/pull/98659#pullrequestreview-1376822560
   bool materialize_non_diff_grads;
 
+  // When true, PyNode::apply uses boxed calling convention: grad tensors are
+  // moved from the immutable args tuple into a mutable list (_boxed_grads) so
+  // the backward can release individual tensor references mid-execution.
```
Comment (albanD) [inline torch/csrc/autograd/python_function.h]: Can you add a more concrete comment here that explains what is the new calling convention when this is enabled since there are no other docs.
This doesn't need to mention a special case on how this can be used.

File: torch/csrc/autograd/python_function.cpp [hunk 9]
```diff
@@ -160,11 +160,37 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
...
+    THPObjectPtr gradsList(PyList_New(n));
+    if (!gradsList)
+      throw_python_error();
+    for (Py_ssize_t i = 0; i < n; i++) {
+      PyObject* item = PyTuple_GET_ITEM(pyInputs.get(), i);
+      Py_INCREF(item);
+      PyList_SET_ITEM(gradsList.get(), i, item);
+      Py_INCREF(Py_None);
+      Py_DECREF(PyTuple_GET_ITEM(pyInputs.get(), i));
+      PyTuple_SET_ITEM(pyInputs.get(), i, Py_None);
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp]: This is prone to race condition. You need to remove the item before decref.

File: torch/csrc/autograd/python_function.cpp [hunk 10]
```diff
@@ -160,11 +160,37 @@ auto PyNode::apply(variable_list&& inputs) -> variable_list {
...
+    }
+    if (PyObject_SetAttrString(obj, "_boxed_grads", gradsList.get()) < 0)
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp]: Why this?
I was expecting that you would get backward(ctx, stack) -> None as the backward function?!

File: torch/csrc/autograd/python_function.cpp [hunk 11]
```diff
@@ -1510,6 +1537,31 @@ int THPFunction_set_materialize_grads(
   END_HANDLE_TH_ERRORS_RET(-1)
 }
 
+PyObject* THPFunction_get_boxed_grads_call(THPFunction* self, void* _unused) {
```
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp]: Can you make these class attributes rather than this.
More similar to `clear_saved_tensors_on_access`.

File: test/test_autograd.py [hunk 12]
```diff
@@ -8665,6 +8665,184 @@ def backward(ctx, g):
         self.assertEqual(y.grad_fn.saved_tensors, ())
         self.assertEqual(y.grad_fn._raw_saved_tensors, ())
 
+    @unittest.skipIf(not TEST_CUDA, "CUDA is unavailable")
+    def test_custom_function_boxed_grads(self):
+        """Test the C++ boxed_grads_call mechanism without torch.compile.
```
Comment (albanD) [inline test/test_autograd.py]: Why c++?

File: test/test_autograd.py [hunk 13]
```diff
@@ -8665,6 +8665,184 @@ def backward(ctx, g):
...
+        # d/dx (2x + 3x).sum() = 2*2 + 3*3 = 13
+        # forward: a=2x, b=3x; backward: grad_a=1, grad_b=1
+        # return grad_a*2 + grad_b*3 = 2+3 = 5
+        self.assertEqual(x.grad, torch.full_like(x, 5.0))
+
+    @unittest.skipIf(not TEST_CUDA, "CUDA is unavailable")
+    def test_custom_function_boxed_grads_no_extra_refs(self):
+        """Framework holds no extra refs to grads with boxed calling convention.
+
+        sys.getrefcount(grad) == 2 inside backward means only the local
```
Comment (albanD) [inline test/test_autograd.py]: We usually use weakref rather than getrefcount to do this.

File: torch/csrc/autograd/python_function.cpp (hunk 9 above)
Comment (IvanKobzarev) [reply torch/csrc/autograd/python_function.cpp]: Thanks! goot catch.

File: test/test_autograd.py (hunk 13 above)
Comment (IvanKobzarev) [reply test/test_autograd.py]: Oh, agree, weakref will be easier and more stable.

File: torch/csrc/autograd/python_function.cpp (hunk 10 above)
Comment (IvanKobzarev) [reply torch/csrc/autograd/python_function.cpp]: Yes, that was my first idea to pass it through a separate attr. True "boxed" function will be cleaner and more straightforward and aligned with "boxed" in python part - I will change to it.

File: torch/csrc/autograd/python_function.h [hunk 14]
```diff
@@ -120,6 +120,19 @@ struct THPFunction {
...
+  // When true, PyNode::apply passes grads as a single mutable list argument
+  // instead of individual args in an immutable tuple, allowing backward to
+  // free grads mid-execution.
+  // Example when it is needed:
+  // pt2 compilation makes AutogradFunction to run compiled forward and compiled
+  // backward, grads are inputs of compiled backward and standard calling
+  // convention holds reference to the grads (immutable args tuple), which
+  // prevents their deallocation after last use in the middle of backward.
+  // This can contribute to the peak memory.
+  // "boxed_grads_call" calling convention allows to pass grads in mutable list
```
Comment (albanD) [inline torch/csrc/autograd/python_function.h]: This duplicates the top comment?

File: torch/csrc/autograd/python_function.h [hunk 15]
```diff
@@ -120,6 +120,19 @@ struct THPFunction {
   // https://github.com/pytorch/pytorch/pull/98659#pullrequestreview-1376822560
   bool materialize_non_diff_grads;
 
+  // When true, PyNode::apply passes grads as a single mutable list argument
```
Comment (albanD) [inline torch/csrc/autograd/python_function.h:123]: The calling convention is still not documented?

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 16]
```diff
@@ -2773,6 +2780,13 @@ def forward(ctx: Any, *deduped_flat_tensor_args: Any) -> Any:
 
             @staticmethod
             def backward(ctx: Any, *flat_args: Any) -> tuple[Any, ...]:
+                # With boxed_grads_call, PyNode::apply passes a single mutable
+                # list as the argument. Otherwise, grads come as *flat_args.
+                if len(flat_args) == 1 and isinstance(flat_args[0], list):
```
Comment (albanD) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: Why do you need this? The flag is always there so you're guaranteed to always get the arguments this way

**Review by albanD — COMMENTED** [review]
Some comments are the same as the previous review?

File: torch/csrc/autograd/python_function.h (hunk 14 above)
Comment (IvanKobzarev) [reply torch/csrc/autograd/python_function.h]: I wrote it myself and get some repetetive wording :) Will try to remove duplication.

File: torch/csrc/autograd/python_function.h (hunk 15 above)
Comment (IvanKobzarev) [reply torch/csrc/autograd/python_function.h:123]: Sorry, did not get it from the first review pass, I thought that you mentioned better comments :) Added documentation.

File: docs/source/notes/extending.rst [hunk 17]
```diff
@@ -153,6 +153,37 @@ class attribute:
   needed once. The default is ``False``. Note that ``saved_tensors`` can only be
   accessed once when this is enabled; a second access will raise an error.
 
+- :attr:`~Function.boxed_grads_call`: When set to ``True`` on the
+  :class:`~Function` subclass, backward receives grads as a single mutable list
+  argument instead of individual args in an immutable tuple. This allows backward
+  to free individual grads mid-execution by removing them from the list, reducing
+  peak memory. When enabled, the backward calling convention changes from
+  ``backward(ctx, *grads)`` to ``backward(ctx, grads)`` where ``grads`` is a list.
```
Comment (albanD) [inline docs/source/notes/extending.rst]: Can you add the return types?

File: docs/source/notes/extending.rst [hunk 18]
```diff
@@ -153,6 +153,37 @@ class attribute:
...
+
+        @staticmethod
+        def forward(ctx, x, weight):
+            ctx.save_for_backward(x, weight)
+            return x.mm(weight)
+
+        @staticmethod
+        def backward(ctx, grads):
+            grad_output = grads[0]
+            grads[0] = None  # Free grad_output reference from the list
```
Comment (albanD) [inline docs/source/notes/extending.rst]:
```suggestion
            grads[0] = None  # Free grad_output reference from the caller
```

File: docs/source/notes/extending.rst [hunk 19]
```diff
@@ -153,6 +153,37 @@ class attribute:
...
+
+            x, weight = ctx.saved_tensors
+            grad_x = grad_output.mm(weight.t())
+            grad_weight = x.t().mm(grad_output)
+            del grad_output  # Now grad_output can be freed
+
```
Comment (albanD) [inline docs/source/notes/extending.rst]: This is a bit of a weird example given that this doesn't actually give any benefit?
Any better example that does have benefit?

Comment (IvanKobzarev) [reply docs/source/notes/extending.rst]: Ok, will replace with QKV Projection example, where we can free grads earlier and actually reduce peak

File: test/test_autograd.py [hunk 20]
```diff
@@ -8665,6 +8665,267 @@ def backward(ctx, g):
...
+                return x * 2
+
+            @staticmethod
+            def backward(ctx, grads):
+                grad = grads[0]
+                grad_for_compute = grad.clone()
+                ref = weakref.ref(grad)
+                grads.clear()
+                del grad
+                result_box["ref_dead"] = ref() is None
```
Comment (albanD) [inline test/test_autograd.py]: You can self.assertTrue() here directly. No need for the nonlocal

File: test/test_autograd.py [hunk 21]
```diff
@@ -8665,6 +8665,267 @@ def backward(ctx, g):
...
+
+        x = torch.randn(4, requires_grad=True)
+        out = Mul3.apply(Mul2.apply(x))
+        out.sum().backward()
+        # backward order: Mul3 then Mul2
+        self.assertEqual(call_order, ["Mul3", "Mul2"])
+        # d/dx (3 * 2 * x).sum() = grad 1 → Mul3 bwd: 1*3=3 → Mul2 bwd: 3*2=6
+        self.assertEqual(x.grad, torch.full_like(x, 6.0))
+
+    def test_custom_function_boxed_grads_none_grads(self):
```
Comment (albanD) [inline test/test_autograd.py:8848]: I guess this also works smoothly with materialize_grads(True) ?

File: docs/source/notes/extending.rst [hunk 22]
```diff
@@ -153,6 +153,49 @@ class attribute:
   needed once. The default is ``False``. Note that ``saved_tensors`` can only be
   accessed once when this is enabled; a second access will raise an error.
 
+- :attr:`~Function.boxed_grads_call`: When set to ``True`` on the
+  :class:`~Function` subclass, backward receives grads as a single mutable list
+  argument instead of individual args in an immutable tuple. This allows backward
```
Comment (albanD) [inline docs/source/notes/extending.rst:158]: Can you mention here that the return from the backward is unchanged?

**Review by albanD — APPROVED** [review]
That sounds good. Only small nits!

**Comment by albanD** [comment]
@claude do a line by line review looking for PyObject* refcount bugs.

File: test/test_autograd.py (hunk 21 above)
Comment (IvanKobzarev) [reply test/test_autograd.py:8848]: I will add the test.
